src/Agent/NewRelic/Agent/Core/NewRelic.Agent.Core/RuntimeEnvironmentInfo.cs

```diff
@@ -0,0 +1,190 @@
+        {
+            public string Id;
+            public string VersionId;
+        }
+
+        private static string GetOSName()
+        {
+            switch (GetOSPlatform())
+            {
+                case Platform.Windows:
+                    return "Windows";
+                case Platform.Linux:
+                    return GetDistroId() ?? "Linux";
+                case Platform.Darwin:
+                    return "Mac OS X";
+                case Platform.FreeBSD:
+                    return "FreeBSD";
+                default:
+                    return "Unknown";
+            }
+        }
+
+        private static string GetOSVersion()
+        {
+            switch (GetOSPlatform())
+            {
+                case Platform.Windows:
+                    return System.Environment.OSVersion.Version.ToString(3);
+                case Platform.Linux:
+                    return GetDistroVersionId() ?? string.Empty;
+                case Platform.Darwin:
+                    return System.Environment.OSVersion.Version.ToString(2);
+                case Platform.FreeBSD:
+                    return GetFreeBSDVersion() ?? string.Empty;
+                default:
+                    return string.Empty;
+            }
+        }
+
+        private static string GetFreeBSDVersion()
+        {
+            // This is same as sysctl kern.version
+            // FreeBSD 11.0-RELEASE-p1 FreeBSD 11.0-RELEASE-p1 #0 r306420: Thu Sep 29 01:43:23 UTC 2016     root@releng2.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC
+            // What we want is major release as minor releases should be compatible.
+            String version = RuntimeInformation.OSDescription;
+            try
+            {
+                // second token up to first dot
+                return RuntimeInformation.OSDescription.Split()[1].Split('.')[0];
+            }
+            catch
+            {
+            }
+            return string.Empty;
+        }
+
+        private static Platform GetOSPlatform()
+        {
+            return _platform.Value;
+        }
+
+        private static string GetDistroId()
+        {
+            return _distroInfo.Value?.Id;
+        }
+
+        private static string GetDistroVersionId()
+        {
+            return _distroInfo.Value?.VersionId;
+        }
+
+        private static DistroInfo LoadDistroInfo()
+        {
+            DistroInfo result = null;
+
+            // Sample os-release file:
+            //   NAME="Ubuntu"
+            //   VERSION = "14.04.3 LTS, Trusty Tahr"
+            //   ID = ubuntu
+            //   ID_LIKE = debian
+            //   PRETTY_NAME = "Ubuntu 14.04.3 LTS"
+            //   VERSION_ID = "14.04"
+            //   HOME_URL = "http://www.ubuntu.com/"
+            //   SUPPORT_URL = "http://help.ubuntu.com/"
+            //   BUG_REPORT_URL = "http://bugs.launchpad.net/ubuntu/"
+            // We use ID and VERSION_ID
+
+            if (File.Exists("/etc/os-release"))
```

Comment:
Probably want a try/catch around this block?

Reply:
Good idea. I'll add that.